src/app/state/tetris/tetris.state.ts

```diff
@@ -0,0 +1,130 @@
+import { PieceFactory } from '@angular-tetris/factory/piece-factory';
+import { GameState } from '@angular-tetris/interface/game-state';
+import { Piece } from '@angular-tetris/interface/piece/piece';
+import { Speed } from '@angular-tetris/interface/speed';
+import { Tile } from '@angular-tetris/interface/tile/tile';
+import { MatrixUtil } from '@angular-tetris/interface/utils/matrix';
+import { LocalStorageService } from '@angular-tetris/services/local-storage.service';
+import { DestroyRef, Injectable, computed, effect, inject, signal } from '@angular/core';
+
+export interface TetrisState {
+  matrix: Tile[];
+  current: Piece | null;
+  next: Piece;
+  hold: Piece;
+  canHold: boolean;
+  points: number;
+  locked: boolean;
+  sound: boolean;
+  initSpeed: Speed;
+  speed: Speed;
+  initLine: number;
+  clearedLines: number;
+  gameState: GameState;
+  saved: TetrisState;
+  max: number;
+}
+
+const createInitialState = (pieceFactory: PieceFactory): TetrisState => ({
+  matrix: MatrixUtil.getStartBoard(),
+  current: null,
+  next: pieceFactory.getRandomPiece(),
+  hold: pieceFactory.getNonePiece(),
+  canHold: true,
+  points: 0,
+  locked: true,
+  sound: true,
+  initLine: 0,
+  clearedLines: 0,
+  initSpeed: 1,
+  speed: 1,
+  gameState: GameState.Loading,
+  saved: null,
+  max: LocalStorageService.maxPoint
+});
+
+const isObjectShallowEqual = (a: any, b: any) => a == b;
+
+@Injectable({ providedIn: 'root' })
+export class TetrisStateService {
+  private pieceFactory = inject(PieceFactory);
+  private tetrisState = signal<TetrisState>(createInitialState(this.pieceFactory));
+
+  next = computed(() => this.tetrisState().next, {
+    equal: isObjectShallowEqual
+  });
+  hold = computed(() => this.tetrisState().hold, {
+    equal: isObjectShallowEqual
+  });
+  matrix = computed(() => this.tetrisState().matrix, {
+    equal: isObjectShallowEqual
+  });
+  current = computed(() => this.tetrisState().current, {
+    equal: isObjectShallowEqual
+  });
+  isEnableSound = computed(() => this.tetrisState().sound);
+  gameState = computed(() => this.tetrisState().gameState);
+  hasCurrent = computed(() => !!this.current());
+  points = computed(() => this.tetrisState().points);
+  clearedLines = computed(() => this.tetrisState().clearedLines);
+  initLine = computed(() => this.tetrisState().initLine);
+  speed = computed(() => this.tetrisState().speed);
+  initSpeed = computed(() => this.tetrisState().initSpeed);
+  max = computed(() => this.tetrisState().max);
+  canStartGame = computed(() => this.gameState() !== GameState.Started);
+  isPlaying = computed(() => this.gameState() === GameState.Started);
+  isPause = computed(() => this.gameState() === GameState.Paused);
+  locked = computed(() => this.tetrisState().locked);
+  canHold = computed(() => this.tetrisState().canHold);
+
+  isShowLogo = signal(false);
```

Comment:
In my opinion, for certain implementations with RxJS that feel more natural, it would be better to stick with RxJS instead of trying to convert everything to signals. By keeping the old RxJS implementation and using `toObservable` to convert a signal into an Observable, the code becomes much more readable compared to using `setTimeout`, `clearTimeout`, and similar methods.

Read more from this [Converting Signals to Observables in Angular ↗](https://netbasal.com/converting-signals-to-observables-in-angular-what-you-need-to-know-971eacd3af2)

```ts
import { toObservable } from '@angular/core/rxjs-interop';

  isShowLogo$ = combineLatest([toObservable(this.gameState), toObservable(this.current)]).pipe(
    switchMap(([state, current]) => {
      const isLoadingOrOver = state === GameState.Loading || state === GameState.Over;
      const isRenderingLogo$ = of(isLoadingOrOver && !current);
      return isLoadingOrOver ? isRenderingLogo$.pipe(delay(1800)) : isRenderingLogo$;
    })
  );
```

![image](https://github.com/trungk18/angular-tetris/assets/6767322/a0e1ed64-65ef-45ff-9973-4332f87c5668)